File: src/main/java/com/natia/secretmod/networking/ColorText.java
package com.natia.secretmod.networking;

import com.natia.secretmod.config.SkyTweaksConfig;
import net.minecraft.util.ChatComponentText;
import net.minecraft.util.EnumChatFormatting;
import net.minecraft.util.IChatComponent;
import net.minecraft.util.StringUtils;
import net.minecraftforge.client.event.ClientChatReceivedEvent;
import net.minecraftforge.event.entity.player.ItemTooltipEvent;
import net.minecraftforge.event.entity.player.PlayerEvent;
import net.minecraftforge.fml.common.eventhandler.EventPriority;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.regex.Pattern;

import static com.natia.secretmod.utils.WebUtils.fetch;

public class ColorText {
    private HashMap<String, String> cache = new HashMap<>();
    private List<String> supporters = new ArrayList<>();
    public ColorText() {
        /* caching & supporters! */
        fetch("https://secretmod-hypixel.herokuapp.com/userswithcolor", res -> {
            System.out.println("List of usernames who bought color cosmetic:");
            for (String supporter: res.asJson().get("user").getAsString().split(Pattern.quote("|"))) {
                if (supporter.equals("")) continue;
                System.out.println("- " + supporter);
                supporters.add(supporter);
                if (!cache.containsKey(supporter)) {
                    fetch("https://secretmod-hypixel.herokuapp.com/usercolor?name=" + supporter, res1 -> {
                        if (res1 != null) {
                            cache.put(supporter, res1.asString());
                        }
                    });
                }
            }
        });
    }

    @SubscribeEvent(priority = EventPriority.LOWEST)
    public void changeText(ClientChatReceivedEvent event) {
        if (!SkyTweaksConfig.colorCosmetic) return;
        if (event.type == 0) {
            IChatComponent sentComponent = event.message;

            if (!event.message.getUnformattedText().contains("-----------------")) {
                String formattedText = sentComponent.getFormattedText();

                for (String name : supporters) {
                    if (event.message.getUnformattedText().contains(name)) {
                        sentComponent = getColor(formattedText, name);
                        formattedText = sentComponent.getFormattedText();
                    }
                }

                event.message = sentComponent;
            }
        }
    }

    @SubscribeEvent(priority = EventPriority.LOWEST)
    public void modifyItemTooltip(ItemTooltipEvent event) {
        if (!SkyTweaksConfig.colorCosmetic) return;
        for (String tip : event.toolTip) {
            String formattedTip = tip;
            for (String name : supporters) {
                if (StringUtils.stripControlCodes(tip).contains(name)) {
                    formattedTip = getColor(formattedTip, name).getFormattedText();
                }
            }
            event.toolTip.set(event.toolTip.indexOf(tip), formattedTip);

        }
    }

    private ChatComponentText getColor(String formattedText, String name) {
        switch (cache.get(name)) {
            case "gold":
                return new ChatComponentText(formattedText.replace(name, EnumChatFormatting.GOLD + "" + name + EnumChatFormatting.RESET));

            case "black":
                return new ChatComponentText(formattedText.replace(name,EnumChatFormatting.BLACK + "" + name + EnumChatFormatting.RESET));

            case "blue":
                return new ChatComponentText(formattedText.replace(name,EnumChatFormatting.BLUE + "" + name + EnumChatFormatting.RESET));

            case "purple":
                return new ChatComponentText(formattedText.replace(name,EnumChatFormatting.DARK_PURPLE + "" + name + EnumChatFormatting.RESET));

            case "red":
                return new ChatComponentText(formattedText.replace(name,EnumChatFormatting.RED + "" + name + EnumChatFormatting.RESET));
        }

        return new ChatComponentText(formattedText);
    }

    @SubscribeEvent(priority = EventPriority.HIGHEST)
    public void renderTags(PlayerEvent.NameFormat event) {
        if (!SkyTweaksConfig.colorCosmetic) return;

        if (supporters == null) return;

        String formattedText = event.username;
        for (String name : supporters) {
            if (event.username.contains(name)) {
                event.displayname = getColor(formattedText, name).getFormattedText();
                break;
            }
        }
    }
}

File: src/main/kotlin/natia/skytweaks/networking/ColorText.kt
package natia.skytweaks.networking

import natia.skytweaks.SkyTweaks
import mixin.natia.skytweaks.SkyTweaksConfig
import net.minecraft.entity.EntityLivingBase
import net.minecraft.util.ChatComponentText
import net.minecraft.util.EnumChatFormatting
import net.minecraft.util.IChatComponent
import net.minecraft.util.StringUtils
import net.minecraftforge.client.event.ClientChatReceivedEvent
import net.minecraftforge.client.event.RenderLivingEvent
import net.minecraftforge.event.entity.player.ItemTooltipEvent
import net.minecraftforge.event.entity.player.PlayerEvent
import net.minecraftforge.fml.common.eventhandler.EventPriority
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent

import java.util.ArrayList
import java.util.HashMap
import java.util.regex.Pattern

import natia.skytweaks.utils.WebUtils.fetch

class ColorText {
    private val cache = HashMap<String, String>()
    private val supporters = ArrayList<String>()

    init {
        /* caching & supporters! */
        fetch("https://skytweaks-web.vercel.app/userswithcolor") { res ->
            SkyTweaks.LOGGER.info("List of usernames who bought color cosmetic:")
            for (supporter in res.asJson().get("user").asString.split("|")) {
                if (supporter == "") continue
                SkyTweaks.LOGGER.info("- $supporter")
                supporters.add(supporter)
                if (!cache.containsKey(supporter)) {
                    fetch("https://skytweaks-web.vercel.app/usercolor?name=$supporter") { res1 ->
                        cache[supporter] = res1.asString()
                    }
                }
            }
        }
    }

    @SubscribeEvent(priority = EventPriority.LOWEST)
    fun onLivingRender(event: RenderLivingEvent.Specials.Pre<EntityLivingBase>) {
        if (SkyTweaksConfig.colorCosmetic) return

        val entity = event.entity
        if (entity.hasCustomName()) {
            for (name in supporters) {
                if (entity.customNameTag.contains(name)) {
                    entity.customNameTag = getColor(entity.customNameTag, name).formattedText
                }
            }
        }
    }

    @SubscribeEvent(priority = EventPriority.LOWEST)
    fun changeText(event: ClientChatReceivedEvent) {
        if (!SkyTweaksConfig.colorCosmetic) return
        if (event.type.toInt() == 0) {
            var sentComponent = event.message

            if (!event.message.unformattedText.contains("-----------------")) {
                for (name in supporters) {
                    if (event.message.unformattedText.contains(name)) {
                        sentComponent = replaceChat(sentComponent, name)
                    }
                }
                event.message = sentComponent
            }
        }
    }

    @SubscribeEvent(priority = EventPriority.LOWEST)
    fun modifyItemTooltip(event: ItemTooltipEvent) {
        if (!SkyTweaksConfig.colorCosmetic) return
        for (tip in event.toolTip) {
            var formattedTip = tip
            for (name in supporters) {
                if (StringUtils.stripControlCodes(tip).contains(name)) {
                    formattedTip = getColor(formattedTip, name).formattedText
                }
            }
            event.toolTip[event.toolTip.indexOf(tip)] = formattedTip

        }
    }

    private fun replaceChat(component: IChatComponent, user: String): IChatComponent {
        val newComponent: IChatComponent
        val text = component as ChatComponentText

        newComponent = getColor(text.unformattedTextForChat, user)
        newComponent.setChatStyle(text.chatStyle.createShallowCopy())

        for (sibling in text.siblings) {
            newComponent.appendSibling(replaceChat(sibling, user))
        }

        return newComponent
    }

    private fun getColor(formattedText: String, name: String): ChatComponentText {
        when (cache[name]) {
            "gold" -> return ChatComponentText(formattedText.replace(name, EnumChatFormatting.GOLD.toString() + "" + name + EnumChatFormatting.RESET))

            "black" -> return ChatComponentText(formattedText.replace(name, EnumChatFormatting.BLACK.toString() + "" + name + EnumChatFormatting.RESET))

            "blue" -> return ChatComponentText(formattedText.replace(name, EnumChatFormatting.BLUE.toString() + "" + name + EnumChatFormatting.RESET))

            "purple" -> return ChatComponentText(formattedText.replace(name, EnumChatFormatting.DARK_PURPLE.toString() + "" + name + EnumChatFormatting.RESET))

            "red" -> return ChatComponentText(formattedText.replace(name, EnumChatFormatting.RED.toString() + "" + name + EnumChatFormatting.RESET))
        }

        return ChatComponentText(formattedText)
    }

    @SubscribeEvent(priority = EventPriority.HIGHEST)
    fun renderTags(event: PlayerEvent.NameFormat) {
        if (!SkyTweaksConfig.colorCosmetic) return

        if (supporters == null) return

        val formattedText = event.username
        for (name in supporters) {
            if (event.username.contains(name)) {
                event.displayname = getColor(formattedText, name).formattedText
                break
            }
        }
    }
}
